lang/src/org/partiql/lang/eval/physical/RelationThunk.kt

```diff
@@ -0,0 +1,44 @@
+package org.partiql.lang.eval.physical
+
+import com.amazon.ionelement.api.MetaContainer
+import org.partiql.lang.ast.SourceLocationMeta
+import org.partiql.lang.errors.ErrorCode
+import org.partiql.lang.errors.Property
+import org.partiql.lang.eval.EvaluationException
+import org.partiql.lang.eval.errorContextFrom
+import org.partiql.lang.eval.fillErrorContext
+import org.partiql.lang.eval.relation.RelationIterator
+
+/** A thunk that returns a [RelationIterator], which is the result of evaluating a relational operator. */
+internal typealias RelationThunkEnv = (EvaluatorState) -> RelationIterator
+
+/**
+ * Invokes [t] with error handling like is supplied by [org.partiql.lang.eval.ThunkFactory].
+ *
+ * This function is not currently in `ThunkFactory` to avoid complicating `ThunkFactory` further.  If a need arises,
+ * it could be moved to `ThunkFactory`.
+ */
+internal inline fun relationThunk(metas: MetaContainer, crossinline t: RelationThunkEnv): RelationThunkEnv {
+    val sourceLocationMeta = metas[SourceLocationMeta.TAG] as? SourceLocationMeta
+    return { env: EvaluatorState ->
+        try {
+            t(env)
+        } catch (e: EvaluationException) {
+            // Only add source location data to the error context if it doesn't already exist
+            // in [errorContext].
+            if (!e.errorContext.hasProperty(Property.LINE_NUMBER)) {
+                sourceLocationMeta?.let { fillErrorContext(e.errorContext, sourceLocationMeta) }
+            }
+            throw e
+        } catch (e: Exception) {
+            val message = e.message ?: "<NO MESSAGE>"
+            throw EvaluationException(
+                "Generic exception, $message",
```

Comment:
This `EVALUATOR_GENERIC_EXCEPTION` is the code assigned by `DEFAULT_EXCEPTION_HANDLER_FOR_LEGACY_MODE` when the exception being rethrown is not an `EvaluationException` and therefore does not already have an error code.  

Since this is the same situation here, I don't see a reason why it would need a different error code.